jinahub/indexers/searcher/FaissSearcher/faiss_searcher.py

```diff
@@ -166,76 +162,96 @@ def _build_index(self, vecs_iter: Iterable['np.ndarray']):
                 'Invalid distance metric for Faiss index construction. Defaulting to l2 distance'
             )
 
-        index = self.to_device(
-            index=faiss.index_factory(self.num_dim, self.index_key, metric)
-        )
+        if self.trained_index_file and os.path.exists(self.trained_index_file):
+            index = faiss.read_index(self.trained_index_file)
+            assert index.metric_type == metric
+            assert index.ntotal == 0
+            assert index.d == self.num_dim
+            assert index.is_trained
+        else:
+            index = faiss.index_factory(self.num_dim, self.index_key, metric)
+
+        index.nprobe = self.nprobe
+
+        self.index = self.to_device(index)
 
-        if self.requires_training:
-            if self.train_filepath:
-                train_data = self._load_training_data(self.train_filepath)
 
+    def _build_index(self, vecs_iter: Iterable['np.ndarray']):
```

Comment:
I think `Iterable` is a too weak requirement here, since you iterate over `vecs_iter` several times.

Reply:
Due to limited RAM, `Iterable` is chosen here to load large-scale data with a streaming method. To make the iteration safe, I iterate over the `vecs_iter` using batch mode (the batch_size is provided by the argument of `prefetch_size`)

Reply:
But you consume `vecs_iter` twice. Once in line 197 or 201 and once in the `_build_partial_index` call. You cannot consume an `Iterable` twice.

```def get_iter(count):
    for i in range(count):
        yield i


my_iter = get_iter(10)

first = list(my_iter)
second = list(my_iter)
print('first: ', first)
print('second: ', second)
```
outputs:
```
first:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
second:  []
```

Reply:
The elements fetching from the first iteration in line 197 are temporally filled into the list `self._prefetch_data`. And before the second iteration happened in the `_build_partial_index`, I first consume the data from `self._prefetch_data`. So, we did not lose data.

Reply:
Ah I see. That makes a whole lot of sense. Thanks for the explanation.